Test log:

It seems, the hang is caused by this.
This is a critical error and should be fixed immediately. The root cause is unknown at the moment. But local test works fine. So it could be orchestration/missing configuration.

```
======================================================================
ERROR: Test generating report with QLR files.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/app/headless/tasks/test/helpers.py", line 69, in wrapper
    return func(*args, **kwargs)
  File "/home/app/headless/tasks/test/test_generate_report.py", line 74, in test_generate_report_qlr
    result = async_result.get()
  File "/usr/local/lib/python2.7/dist-packages/celery/result.py", line 191, in get
    on_message=on_message
  File "/usr/local/lib/python2.7/dist-packages/celery/backends/base.py", line 462, in wait_for_pending
    return result.maybe_throw(propagate=propagate, callback=callback)
  File "/usr/local/lib/python2.7/dist-packages/celery/result.py", line 296, in maybe_throw
    self.throw(value, self._to_remote_traceback(tb))
  File "/usr/local/lib/python2.7/dist-packages/celery/result.py", line 289, in throw
    self.on_ready.throw(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/vine/promises.py", line 217, in throw
    reraise(type(exc), exc, tb)
  File "<string>", line 1, in reraise
Exception: The computed count of output layers is wrong. It should be 5 but the count is 6.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -7,17 +7,21 @@ python:
 
 env:
   global:
-    - COMPOSE_PROJECT_NAME: headless
+    - COMPOSE_PROJECT_NAME=headless
   matrix:
-    - ON_TRAVIS: 'True'
-      TASK_ALWAYS_EAGER: 'True'
-    - ON_TRAVIS: 'True'
-      TASK_ALWAYS_EAGER: 'False'
-      WORKER_SCALE: 4
+    - ON_TRAVIS=True
+      TASK_ALWAYS_EAGER=True
+      HEADLESS_COMMAND=dev
+    - ON_TRAVIS=True
+      TASK_ALWAYS_EAGER=False
+      HEADLESS_COMMAND="coverage run --rcfile=/home/app/.coveragerc /usr/local/bin/celery -A headless.celery_app worker -l info -Q inasafe-headless -n inasafe-headless.%h -P solo"
+      WORKER_SCALE=1
+      INASAFE_LOG_LEVEL=10 # DEBUG log level
     # Special environment for subsequent run stress testing
-    - ON_TRAVIS: 'True'
-      TASK_ALWAYS_EAGER: 'True'
-      SUBSEQUENT_RUN_TESTING: 'True'
+    - ON_TRAVIS=True
+      TASK_ALWAYS_EAGER=True
+      HEADLESS_COMMAND=dev
+      SUBSEQUENT_RUN_TESTING=True
 
 virtualenv:
   system_site_packages: true
@@ -50,8 +54,7 @@ script:
   - pushd deployment
   - make check-flake8
   - ../src/scripts/start_coverage_unittest_worker.sh
-  - export CURRENT_DIR=${PWD}
-  - pushd ../src; scripts/run_coverage_tests.sh ${CURRENT_DIR}; popd
+  - ../src/scripts/run_coverage_tests.sh ${PWD} ../src
   - ../src/scripts/stop_coverage_unittest_worker.sh || true
   - popd
 

--- a/deployment/ansible/development/group_vars/all.travis.yml
+++ b/deployment/ansible/development/group_vars/all.travis.yml
@@ -20,7 +20,7 @@ inasafe:
   depth: 1
 
 inasafe_headless_worker:
-  command: dev
+  command: "${HEADLESS_COMMAND}"
   environment:
     TASK_ALWAYS_EAGER: "${TASK_ALWAYS_EAGER}"
     ON_TRAVIS: "${ON_TRAVIS}"
@@ -29,7 +29,8 @@ inasafe_headless_worker:
 
 docker_port_forward:
   inasafe_headless:
-    ssh: *inasafe_headless_ssh_port
+    # Do not port forward, to allow scaling
+    non_ssh: *inasafe_headless_ssh_port
 
 integration_test:
   # Path to inasafe-headless's test data directory

--- a/deployment/ansible/development/roles/docker_compose/templates/docker-compose.override.yml.j2
+++ b/deployment/ansible/development/roles/docker_compose/templates/docker-compose.override.yml.j2
@@ -5,7 +5,7 @@ services:
 
   inasafe-headless-worker:
     build: docker-headless
-    command: {{ inasafe_headless_worker.command }}
+    command: "{{ inasafe_headless_worker.command }}"
     volumes:
 
       {% if 'custom_path' in inasafe %}
@@ -53,6 +53,10 @@ services:
         {% endfor %}
       {% endif %}
 
+    {% if 'ssh' in docker_port_forward.inasafe_headless %}
+
     ports:
       # expose ssh ports
       - "{{ docker_port_forward.inasafe_headless.ssh }}:22"
+
+    {% endif %}

--- a/src/scripts/run_coverage_tests.sh
+++ b/src/scripts/run_coverage_tests.sh
@@ -7,18 +7,24 @@ echo "Running Headless coverage test"
 echo "COMPOSE_PROJECT_NAME=$COMPOSE_PROJECT_NAME"
 export CURRENT_DIR=$1
 echo "Docker compose config directory: $CURRENT_DIR"
+export PACKAGE_DIR=$2
+echo "Package dir to check: $PACKAGE_DIR"
 
 run_coverage_script () {
 	pushd $CURRENT_DIR
 	if [ -z "${SUBSEQUENT_RUN_TESTING}" ]; then
-		docker-compose exec inasafe-headless-worker /bin/bash -c "coverage run --rcfile=/home/app/.coveragerc /usr/local/bin/nosetests --process-restartworker -v $1";
+		docker-compose run --rm inasafe-headless-worker /bin/bash -c "coverage run --rcfile=/home/app/.coveragerc /usr/local/bin/nosetests --process-restartworker -v $1";
 	else
-		docker-compose exec inasafe-headless-worker /bin/bash -c "coverage run --rcfile=/home/app/.coveragerc /usr/local/bin/nosetests -v headless.tasks.test.test_subsequent_run";
+		docker-compose run --rm inasafe-headless-worker /bin/bash -c "coverage run --rcfile=/home/app/.coveragerc /usr/local/bin/nosetests -v headless.tasks.test.test_subsequent_run";
 	fi
+	exit_code=$?
 	popd
+	return $exit_code
 }
 
 shopt -s globstar
+
+pushd $PACKAGE_DIR
 
 for filename in **/test_*.py; do
 	echo "Test $filename"
@@ -30,10 +36,15 @@ for filename in **/test_*.py; do
 
 		if [ "$exit_code" -eq "1" ]; then
 			echo "Unittests failed"
-			exit 1
+
+			echo "Printing immediate headless celery worker logs"
+			pushd $CURRENT_DIR
+			docker-compose logs --tail=100 inasafe-headless-worker
+			popd
+			break 2
 		elif [ "$exit_code" -eq "139" ]; then
 			echo "Unittests finished but doesn't exit cleanly"
-			exit 0
+			exit_code=0
 		fi
 		# investigate why it failed
 		echo "Is it memory error?"
@@ -44,9 +55,17 @@ for filename in **/test_*.py; do
 
 		# Restart attempt
 		echo "$PWD"
+		pushd $CURRENT_DIR
 		make up
+		docker-compose restart inasafe-headless-worker
+		popd
 		sleep 10
+		if [ "$exit_code" -eq "0" ]; then
+			break
+		fi
 	done
 done
 
+popd
 echo "Testing finished"
+exit $exit_code

--- a/src/scripts/start_coverage_unittest_worker.sh
+++ b/src/scripts/start_coverage_unittest_worker.sh
@@ -3,8 +3,8 @@
 set -e
 
 if [ -z "${TASK_ALWAYS_EAGER}" ] || [ "${TASK_ALWAYS_EAGER}" = "False" ]; then
-	for i in `seq 1 ${WORKER_SCALE}`;
-	do
-		make coverage-worker;
-	done
+	echo "Command: ${HEADLESS_COMMAND}"
+	docker-compose scale inasafe-headless-worker=${WORKER_SCALE}
+	sleep 5
+	docker-compose logs inasafe-headless-worker
 fi

--- a/src/scripts/stop_coverage_unittest_worker.sh
+++ b/src/scripts/stop_coverage_unittest_worker.sh
@@ -1,8 +1,6 @@
 #!/usr/bin/env bash
 
 if [ -z "${TASK_ALWAYS_EAGER}" ] || [ "${TASK_ALWAYS_EAGER}" = "False" ]; then
-	for i in `seq 1 ${WORKER_SCALE}`;
-	do
-		make coverage-worker-kill WORKER_ID=${i};
-	done
+	docker-compose scale inasafe-headless-worker=0
+	sleep 5
 fi

